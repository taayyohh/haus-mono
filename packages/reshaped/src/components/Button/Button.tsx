import React from "react";
import { classNames, responsiveClassNames, responsivePropDependency } from "utilities/helpers";
import Icon from "components/Icon";
import Loader from "components/Loader";
import Actionable, { ActionableRef } from "components/Actionable";
import ButtonAligner from "./ButtonAligner";
import type * as T from "./Button.types";
import s from "./Button.module.css";

const ButtonBase = (props: T.Props, ref: ActionableRef) => {
	const {
		variant = "solid",
		color = "neutral",
		elevated,
		highlighted,
		fullWidth,
		loading,
		disabled,
		type,
		href,
		size = "medium",
		className,
		children,
		rounded,
		attributes,
		onClick,
		icon,
		endIcon,
	} = props;
	const iconOnly = (icon || endIcon) && !children;
	const rootClassName = classNames(
		s.root,
		className,
		color && s[`--color-${color}`],
		variant && s[`--variant-${variant}`],
		responsiveClassNames(s, "--size", size),
		responsiveClassNames(s, "--full-width", fullWidth),
		elevated && variant !== "ghost" && s["--elevated"],
		rounded && s["--rounded"],
		disabled && s["--disabled"],
		loading && s["--loading"],
		highlighted && s["--highlighted"],
		iconOnly && s["--icon-only"]
	);

	const renderIcon = (position: "start" | "end") => {
		const isStartValid = position === "start" && icon;
		const isEndValid = position === "end" && endIcon;
		const isInvalid = !(isStartValid || isEndValid);

		if (isInvalid) return null;

		const iconClassName = classNames(s.icon, position === "end" && s["--icon-position-end"]);
		const iconSize = responsivePropDependency(size, (size) => {
			if (size === "large") return 5;
			if (size === "xlarge") return 6;
			return 4;
		});

		return (
			<Icon
				className={iconClassName}
				svg={(position === "start" ? icon : endIcon)!}
				size={iconSize}
				autoWidth
			/>
		);
	};

	return (
		<Actionable
			disabled={disabled || loading}
			className={rootClassName}
			attributes={attributes}
			type={type}
			onClick={onClick}
			href={href}
			ref={ref}
		>
			{loading && (
				<div className={s.loader}>
					<Loader size="small" color="inherit" />
				</div>
			)}
			{renderIcon("start")}
			{children && <span className={s.text}>{children}</span>}
			{renderIcon("end")}
		</Actionable>
	);
};

const Button = React.forwardRef(ButtonBase) as T.Export;
Button.Aligner = ButtonAligner;

export default Button;
