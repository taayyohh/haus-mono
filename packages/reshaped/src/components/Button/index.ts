export { default } from "./Button";
export type { Props as ButtonProps, AlignerProps as ButtonAlignerProps } from "./Button.types";
