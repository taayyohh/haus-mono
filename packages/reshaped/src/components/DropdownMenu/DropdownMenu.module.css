.menu {
  padding: var(--rs-unit-x1);
}

.section + .section {
  margin-top: var(--rs-unit-x1);
  padding-top: var(--rs-unit-x1);
  border-top: 1px solid var(--rs-color-border-neutral-faded);
}

.item + .item {
  margin-top: var(--rs-unit-x1);
}

[dir='rtl'] .arrow {
  transform: scaleX(-1);
}
