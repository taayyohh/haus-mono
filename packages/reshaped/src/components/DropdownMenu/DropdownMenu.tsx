"use client";

import React from "react";
import Popover from "components/Popover";
import MenuItem, { MenuItemProps } from "components/MenuItem";
import Icon from "components/Icon";
import { useFlyoutContext, FlyoutInstanceRef } from "components/_private/Flyout";
import IconChevronRight from "icons/ChevronRight";
import useHotkeys from "hooks/useHotkeys";
import useRTL from "hooks/useRTL";
import * as keys from "constants/keys";
import type * as T from "./DropdownMenu.types";
import s from "./DropdownMenu.module.css";

const DropdownMenuSubContext = React.createContext<React.RefObject<FlyoutInstanceRef> | null>(null);

const DropdownMenu = (props: T.Props) => {
	const { children, position = "bottom-start", triggerType = "click", ...popoverProps } = props;

	return (
		<Popover
			{...popoverProps}
			position={position}
			padding={0}
			trapFocusMode="action-menu"
			triggerType={triggerType}
		>
			{children}
		</Popover>
	);
};

const DropdownMenuContent = (props: T.ContentProps) => {
	const { children } = props;
	const subMenuInstance = React.useContext(DropdownMenuSubContext);
	const [rtl] = useRTL();
	const { ref } = useHotkeys<HTMLDivElement>(
		{
			[rtl ? keys.RIGHT : keys.LEFT]: () => {
				subMenuInstance?.current?.close();
			},
		},
		[subMenuInstance?.current]
	);

	return (
		<Popover.Content attributes={{ ref }}>
			<div className={s.menu} role="menu">
				{children}
			</div>
		</Popover.Content>
	);
};

const DropdownMenuSection = (props: T.SectionProps) => {
	const { children } = props;

	return (
		<div className={s.section} role="group">
			{children}
		</div>
	);
};

const DropdownMenuItem = (props: Omit<MenuItemProps, "roundedCorners">) => {
	const { onClick } = props;
	const { handleClose } = useFlyoutContext();

	const handleClick = (e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>) => {
		if (handleClose) handleClose({ closeParents: true });
		if (onClick) onClick(e);
	};

	return (
		<MenuItem
			{...props}
			roundedCorners
			className={s.item}
			attributes={{ ...props.attributes, role: "menuitem" }}
			onClick={handleClick}
		/>
	);
};

const DropdownMenuSubMenu = (props: T.SubMenuProps) => {
	const { children } = props;
	const dropdownMenuRef = React.useRef<FlyoutInstanceRef>();

	return (
		<DropdownMenuSubContext.Provider value={dropdownMenuRef}>
			<DropdownMenu
				triggerType="hover"
				position="end-top"
				contentGap={0.5}
				instanceRef={dropdownMenuRef}
			>
				{children}
			</DropdownMenu>
		</DropdownMenuSubContext.Provider>
	);
};

const DropdownMenuSubTriggerItem = (props: Omit<MenuItemProps, "endSlot" | "roundedCorners">) => {
	const { children, attributes, ...menuItemProps } = props;
	const subMenuInstance = React.useContext(DropdownMenuSubContext);
	const [rtl] = useRTL();
	const { ref } = useHotkeys(
		{
			[rtl ? keys.LEFT : keys.RIGHT]: () => {
				subMenuInstance?.current?.open();
			},
		},
		[],
		{ ref: attributes?.ref }
	);

	return (
		<DropdownMenuItem
			{...menuItemProps}
			attributes={{ ...attributes, ref }}
			endSlot={<Icon autoWidth svg={IconChevronRight} className={s.arrow} />}
		>
			{children}
		</DropdownMenuItem>
	);
};

const DropdownMenuSubTrigger = (props: T.SubTriggerProps) => {
	const { children } = props;

	return (
		<DropdownMenu.Trigger>
			{(attributes) => (
				<DropdownMenuSubTriggerItem attributes={attributes}>{children}</DropdownMenuSubTriggerItem>
			)}
		</DropdownMenu.Trigger>
	);
};

DropdownMenu.Trigger = Popover.Trigger;
DropdownMenu.Content = DropdownMenuContent;
DropdownMenu.Section = DropdownMenuSection;
DropdownMenu.Item = DropdownMenuItem;
DropdownMenu.SubMenu = DropdownMenuSubMenu;
DropdownMenu.SubTrigger = DropdownMenuSubTrigger;

export default DropdownMenu;
