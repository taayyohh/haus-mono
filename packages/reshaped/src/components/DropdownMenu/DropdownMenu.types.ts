import React from "react";
import type { PopoverProps } from "components/Popover";

export type Props = Pick<
	PopoverProps,
	| "children"
	| "position"
	| "forcePosition"
	| "triggerType"
	| "contentGap"
	| "onOpen"
	| "onClose"
	| "active"
	| "defaultActive"
	| "width"
	| "instanceRef"
>;

export type ContentProps = {
	children: React.ReactNode;
};

export type SectionProps = {
	children: React.ReactNode;
};

export type SubMenuProps = {
	children: React.ReactNode;
};

export type SubTriggerProps = {
	children: React.ReactNode;
};
