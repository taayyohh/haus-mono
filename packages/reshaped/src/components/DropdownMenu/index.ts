export { default } from "./DropdownMenu";
export type { Props as DropdownMenuProps } from "./DropdownMenu.types";
