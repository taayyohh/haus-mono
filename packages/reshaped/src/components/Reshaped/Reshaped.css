:root {
  --rs-z-index-raised: 5;
  --rs-z-index-flyout: 80;
  --rs-z-index-fixed: 90;
  --rs-z-index-overlay: 100;
  --rs-z-index-notification: 110;
}

[data-rs-color-mode='light'] {
  color-scheme: light;
}

[data-rs-color-mode='dark'] {
  color-scheme: dark;
}

* {
  box-sizing: border-box;
}

body,
h1,
h2,
h3,
h4,
h5,
h6,
p,
ul,
ol,
li,
figure,
figcaption,
blockquote,
dl,
dd {
  margin: 0;
  padding: 0;
}

ul[class],
ol[class] {
  list-style: none;
}

a:not([class]) {
  text-decoration-skip-ink: auto;
}

img {
  max-width: 100%;
  display: block;
}

input,
button,
textarea,
select {
  font: inherit;
}

/* 
Setting background for Windows browsers that have white background by default 
which doesn't work well in dark mode
*/
option {
  background: var(--rs-color-background-elevation-base);
}

label {
  cursor: pointer;
}

@media (prefers-reduced-motion: reduce) {
  * {
    animation-duration: 0.01ms !important;
    animation-iteration-count: 1 !important;
    transition-duration: 0.01ms !important;
    scroll-behavior: auto !important;
  }
}

[data-rs-no-transition] *,
[data-rs-no-transition] *::before,
[data-rs-no-transition] *::after {
  transition: none !important;
}

input::placeholder {
  color: var(--rs-color-neutral-faded);
  opacity: 0.5;
}

html {
  height: 100%;

  /* https://developer.mozilla.org/en-US/docs/Web/CSS/text-size-adjust */
  -webkit-text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;

  /* https://developer.mozilla.org/en-US/docs/Web/CSS/text-rendering */
  text-rendering: optimizelegibility;

  /* https://developer.mozilla.org/en-US/docs/Web/CSS/touch-action#values */
  touch-action: manipulation;
}

body {
  height: 100%;
  scroll-behavior: smooth;
  text-rendering: optimizespeed;
  color: var(--rs-color-foreground-neutral);
  background: var(--rs-color-background-page);
}

body,
[data-rs-theme] {
  font-size: var(--rs-font-size-body-3);
  line-height: var(--rs-line-height-body-3);
  font-family: var(--rs-font-family-body);
  font-weight: var(--rs-font-weight-regular);
}

textarea {
  resize: vertical;
}

table {
  border-collapse: collapse;
  border-spacing: 0;
}

fieldset {
  border: 0;
  margin: 0;
  padding: 0;
}

a {
  text-decoration: underline;
  color: var(--rs-color-foreground-primary);

  &:hover {
    text-decoration: none;
  }
}
