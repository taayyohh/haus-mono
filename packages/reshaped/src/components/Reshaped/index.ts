export { default } from "./Reshaped";
export type { Props as ReshapedProps } from "./Reshaped.types";
