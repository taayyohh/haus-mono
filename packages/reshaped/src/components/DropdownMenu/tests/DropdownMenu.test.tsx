import React from "react";
import { vi } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import DropdownMenu from "components/DropdownMenu";
import Button from "components/Button";
import Reshaped from "components/Reshaped";

const fixtures = {
	content1: "Item 1",
	content2: "Item 2",
};

describe("DropdownMenu", () => {
	test("works when opening and closing", async () => {
		const handleOpen = vi.fn();
		const handleClose = vi.fn();

		render(
			<Reshaped>
				<DropdownMenu onOpen={handleOpen} onClose={handleClose}>
					<DropdownMenu.Trigger>
						{(attributes) => <Button attributes={attributes}>Open</Button>}
					</DropdownMenu.Trigger>
					<DropdownMenu.Content>
						<DropdownMenu.Item>{fixtures.content1}</DropdownMenu.Item>
						<DropdownMenu.Item>{fixtures.content2}</DropdownMenu.Item>
					</DropdownMenu.Content>
				</DropdownMenu>
			</Reshaped>
		);

		const button = screen.getByRole("button");

		expect(button).toBeInTheDocument();
		expect(button).toMatchSnapshot();
		expect(button).not.toHaveAttribute("aria-controls");
		expect(screen.queryByRole("menu")).not.toBeInTheDocument();

		await userEvent.click(button);
		expect(button).toMatchSnapshot();

		await waitFor(() => {
			const menu = screen.queryByRole("menu");
			const menuItems = screen.queryAllByRole("menuitem");

			expect(screen.queryByText(fixtures.content1)).toBeInTheDocument();
			expect(menu).toBeInTheDocument();
			expect(menuItems).toHaveLength(2);

			const ariaControls = button.getAttribute("aria-controls");
			const dropdownEl = ariaControls && document.getElementById(ariaControls);
			expect(dropdownEl).toBeInTheDocument();

			// TODO: Add support for trapFocus unit testing
			// expect(menuItems[0]).toHaveFocus();

			expect(handleOpen).toBeCalledTimes(1);
		});

		await userEvent.click(button);

		await waitFor(() => {
			expect(handleClose).toBeCalledTimes(1);
		});

		// TODO: Add support for handling removal animation
		// await waitForElementToBeRemoved(() => screen.queryByText(fixtures.content1));
		// expect(screen.queryByText(fixtures.content1)).not.toBeInTheDocument();
	});

	test("works as controlled", async () => {
		const handleOpen = vi.fn();
		const handleClose = vi.fn();

		render(
			<Reshaped>
				<DropdownMenu onOpen={handleOpen} onClose={handleClose} active>
					<DropdownMenu.Trigger>
						{(attributes) => <Button attributes={attributes}>Open</Button>}
					</DropdownMenu.Trigger>
					<DropdownMenu.Content>
						<DropdownMenu.Item>{fixtures.content1}</DropdownMenu.Item>
						<DropdownMenu.Item>{fixtures.content2}</DropdownMenu.Item>
					</DropdownMenu.Content>
				</DropdownMenu>
			</Reshaped>
		);

		const button = screen.getByRole("button");

		expect(button).toMatchSnapshot();
		expect(handleOpen).toBeCalledTimes(0);

		await userEvent.click(button);
		expect(handleClose).toBeCalledTimes(1);
		// TODO: Add support for handling removal animation
		// expect(button).toMatchSnapshot();
	});

	test("works as uncontrolled", async () => {
		const handleOpen = vi.fn();
		const handleClose = vi.fn();

		render(
			<Reshaped>
				<DropdownMenu onOpen={handleOpen} onClose={handleClose} defaultActive>
					<DropdownMenu.Trigger>
						{(attributes) => <Button attributes={attributes}>Open</Button>}
					</DropdownMenu.Trigger>
					<DropdownMenu.Content>
						<DropdownMenu.Item>{fixtures.content1}</DropdownMenu.Item>
						<DropdownMenu.Item>{fixtures.content2}</DropdownMenu.Item>
					</DropdownMenu.Content>
				</DropdownMenu>
			</Reshaped>
		);

		const button = screen.getByRole("button");

		expect(button).toMatchSnapshot();
		expect(handleOpen).toBeCalledTimes(0);

		await userEvent.click(button);
		expect(handleClose).toBeCalledTimes(1);

		// TODO: Add support for handling removal animation
		// expect(button).toMatchSnapshot();
	});
});
