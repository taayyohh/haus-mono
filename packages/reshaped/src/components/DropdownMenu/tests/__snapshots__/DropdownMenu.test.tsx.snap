// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropdownMenu > works as controlled 1`] = `
<button
  aria-controls=":r2:"
  aria-expanded="true"
  aria-haspopup="menu"
  class=" _root_6c3f4f  _root_2a1efe _--color-neutral_2a1efe _--variant-solid_2a1efe  _--size-medium_2a1efe"
  type="button"
>
  <span
    class="_text_2a1efe"
  >
    Open
  </span>
</button>
`;

exports[`DropdownMenu > works as uncontrolled 1`] = `
<button
  aria-controls=":r4:"
  aria-expanded="true"
  aria-haspopup="menu"
  class=" _root_6c3f4f  _root_2a1efe _--color-neutral_2a1efe _--variant-solid_2a1efe  _--size-medium_2a1efe"
  type="button"
>
  <span
    class="_text_2a1efe"
  >
    Open
  </span>
</button>
`;

exports[`DropdownMenu > works when opening and closing 1`] = `
<button
  aria-expanded="false"
  aria-haspopup="menu"
  class=" _root_6c3f4f  _root_2a1efe _--color-neutral_2a1efe _--variant-solid_2a1efe  _--size-medium_2a1efe"
  type="button"
>
  <span
    class="_text_2a1efe"
  >
    Open
  </span>
</button>
`;

exports[`DropdownMenu > works when opening and closing 2`] = `
<button
  aria-controls=":r0:"
  aria-expanded="true"
  aria-haspopup="menu"
  class=" _root_6c3f4f  _root_2a1efe _--color-neutral_2a1efe _--variant-solid_2a1efe  _--size-medium_2a1efe"
  type="button"
>
  <span
    class="_text_2a1efe"
  >
    Open
  </span>
</button>
`;
